docs/design/pvc-template.md

```diff
@@ -0,0 +1,246 @@
+      API calls per account - if there is a burst of volume creation and cloud provider started to throttle calls, all
+      clusters under the entire cloud account would get affected, and massive back-offs are hard to handle given the
+      error might propagate through upstream services
+      
+### Non-Goals
+- Volume pooling for reducing Create/Delete calls as we can re-use volume for different jobs
+- Smart throttling such that one training job will not starve other ones due to volume throttling, similar for different
+  worker types within same training job
+
+
+## Proposal
+
+### Implementation Details
+
+I suggest that we start to support dynamic volume provisioning from `v1beta2`, and change will not be back ported to older versions.
+Currently, both tf-operator and pytorch-operator need to be changed.
+
+#### API Change
+Similar to stateful set, we will add a field in replica spec to define PVC template:
+```go
+// ReplicaSpec is a description of the replica
+type ReplicaSpec struct {
+	// ...
+	
+	// VolumeClaimTemplates specifies a list of volume claim templates, which defines volumes specified in template.
+	VolumeClaimTemplates []corev1.PersistentVolumeClaim `json:"volumeClaimTemplates,omitempty"`
+}
+```
+
+Since both `TFJob` and `PyTorchJob` will use `common.ReplicaSpec` since `v1beta2`, such change can be easily adopted by
+both workloads. List is used here instead of map since according to Kubernetes convention,
+[lists is preferred over maps](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps).
+In this case, if user redefine volume claim templates with same name but different spec, the actual outcome will be undefined.
+
+
+#### PVC Creation
+VolumeClaimTemplate will be per worker type, which means that all replicas will have same volume configuration. The exact
+volume type will be up to the user through storage class, not our responsibility.
+
+To request a volume defined by VolumeClaimTemplates, one need to specify the volume in `ReplicaSpec.Template.Spec.Containers[].VolumeMounts`,
+and provide the template in `ReplicaSpec.VolumeClaimTemplates` with **the same volume name**.
+
+**NOTE:** Volume defined in `ReplicaSpec.Template.Spec.Volumes` will override volume with same name in
+`ReplicaSpec.VolumeClaimTemplates`
+
+To avoid conflict of volume naming, the PVC created will have format `{volname}-{podname}`. If this combination
+does not fit requirement, i.e. name too long, we will have error creating the volume, and in this case, related Pods will
+stop progress in a way such that they stuck in `Pending` state as scheduler is unable to find the volume for scheduling.
+In such case, event will be recorded for debugging purpose.
+
+Volume's ownerReference will be set to the corresponding `TFJob` or `PyTorchJob`, and once volume is created, we will
```

Comment:
This means we can cascade deletion, correct?

Reply:
After a second thought this would be dangerous, as cascade deletion will result in arbitrary order of deleting pods and volumes. In the case while volume is getting detached from the host while pod is still shutting down, there could be problems. Here are the 2 other options I have in mind:

### Alternative 1
However making volume's owner to the Pod would also result in inefficiencies as if we restart pod (delete and recreate), volume will be deleted and recreated as well, so the job will always need to recover all its work from scratch.

### Alternative 2
Don't make owner reference at all, but add label about the Volume's Pod and training job name, and rely on a GC process to clean up the volume. More specifically, when a training job is deleted (or marked as delete in progress), we send the volume to the worker to throttle deletion, and when the controller starts up, we find out orphaned volumes and GC them as well

I'd prefer alternative 2, any thoughts from your end?

Reply:
To amend my comments about cascade deletion - I did some experiments and found out that although kubernetes will mark both pod and volume as "termination in progress" at the same time, if they point to the same controller, kubernetes will not clean up volume if the pod is not deleted. so Yes, we are safe to do cascade deletion

Reply:
@zhan849 Are you sure that this is a coding feature instead of a random behaviour? I am not sure if the garbage collector has such a logic.

Reply:
@gaocegege Kubelet is designed to have such protection.

The garbage collector does not have such logic, it will simply mark both objects as "delete in progress", i.e. add a deletion timestamp. However, kubelet has such logic to protect the disk - it will not clean up the disk until the pod is deleted and all containers are deleted. More specifically:

DesiredStateOfWorld will hold off volume deletion when there is still container not yet been terminated.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L225


Volume Reconciler will only start its mount / detach routine once the volume is removed from desired state:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L273

Reply:
Thanks for the links. I am not very familiar with the implementation of Kubelet. Then LGTM